### PR TITLE
[MODINVSTOR-736] Add Related instances relationship type reference record API.

### PR DESCRIPTION
**Pull request by kaladay**
resolves: [MODINVSTOR-736](https://issues.folio.org/browse/MODINVSTOR-736).

Implement "related-instance-type" (not to be confused with "instance-relationship-type") to represent alternate forms or versions of some work.

The following reference date is created:
- Available in another form
- Constituent unit
- Data source
- Has subseries 
- Has supplements
- In
- Issued with
- Main series
- Other edition available
- Related item
- Supplement to
- Translated as
- Translation

### CODE REVIEW

File: src/test/java/org/folio/rest/api/RelatedInstanceTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+import org.folio.rest.support.AdditionalHttpStatusCodes;
+import org.folio.rest.support.Response;
+import org.folio.rest.support.ResponseHandler;
+import org.junit.Before;
+import org.junit.Test;
+
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.json.JsonObject;
+
+public class RelatedInstanceTypeTest extends TestBaseWithInventoryUtil {
```
[the remark this answers is not in the document]
Comment (kaladay) [reply src/test/java/org/folio/rest/api/RelatedInstanceTypeTest.java:26]: I'm not sure why these aren't being covered and I will be investigating this.

File: src/test/java/org/folio/rest/api/RelatedInstanceTypeTest.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+    JsonObject updateRequest = new JsonObject()
+      .put("id", id.toString())
+      .put("name", "Re-enactment of");
+
+    CompletableFuture<Response> updated = new CompletableFuture<>();
+
+    send(relatedInstanceTypesStorageUrl("/" + id.toString()).toString(), HttpMethod.PUT,
+      updateRequest.toString(), SUPPORTED_CONTENT_TYPE_JSON_DEF,
+      ResponseHandler.any(updated));
```
[the remark this answers is not in the document]
Comment (kaladay) [reply src/test/java/org/folio/rest/api/RelatedInstanceTypeTest.java:125]: What I needed to do is virtually identical to existing work.
I based the work exactly off of existing tests, which I believe is MaterialTypeTest.java.

File: src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+  @Validate
+  @Override
+  public void getRelatedInstanceTypes(String query, int offset, int limit, String lang,
+      Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler,
+      Context vertxContext) {
+
+    /**
+     * http://host:port/related-instance-types
+     */
```
[the remark this answers is not in the document]
Comment (kaladay) [reply src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java:54]: Other "type" API's have these.
I don't understand it either, but I opted for conistency.
- https://github.com/folio-org/mod-inventory-storage/blob/8b0f0db1d56cd0563c355cead5af8e0096560f21/src/main/java/org/folio/rest/impl/InstanceTypeAPI.java#L47
- https://github.com/folio-org/mod-inventory-storage/blob/8b0f0db1d56cd0563c355cead5af8e0096560f21/src/main/java/org/folio/rest/impl/ItemNoteTypeAPI.java#L43

Some of them have it in a different spot, like:
- https://github.com/folio-org/mod-inventory-storage/blob/8b0f0db1d56cd0563c355cead5af8e0096560f21/src/main/java/org/folio/rest/impl/CallNumberTypeAPI.java#L40

File: src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                asyncResultHandler.handle(Future.succeededFuture(DeleteRelatedInstanceTypesByRelatedInstanceTypeIdResponse
+                        .respond204()));
+              } catch (Exception e) {
+                internalServerErrorDuringDelete(e, lang, asyncResultHandler);
+              }
+            });
+      } catch (Exception e) {
+        internalServerErrorDuringDelete(e, lang, asyncResultHandler);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (kaladay) [reply src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java:201]: Calling postgres.delete() is the common practice by other "type" APIs, such as:
- https://github.com/folio-org/mod-inventory-storage/blob/8b0f0db1d56cd0563c355cead5af8e0096560f21/src/main/java/org/folio/rest/impl/InstanceRelationshipTypeAPI.java#L167
- https://github.com/folio-org/mod-inventory-storage/blob/8b0f0db1d56cd0563c355cead5af8e0096560f21/src/main/java/org/folio/rest/impl/InstanceTypeAPI.java#L164

I took basing "related instances" off of "instance relationships" quite literally (for the most part).

File: src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                      .respond400WithTextPlain(msg)));
+                }
+              } catch (Exception e) {
+                internalServerErrorDuringPost(e, lang, asyncResultHandler);
+              }
+            });
+      } catch (Exception e) {
+        internalServerErrorDuringPost(e, lang, asyncResultHandler);
+      }
+    });
```
[the remark this answers is not in the document]
Comment (kaladay) [reply src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java:141]: See my response comment about the "delete" PgUtil.

File: src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java:57]: > I don't understand it either, but I opted for conistency.

Personally, I don't think consistency is sufficient to copy something that doesn't provide any value (at least for me)

File: src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java]: > I took basing "related instances" off of "instance relationships" quite literally (for the most part).

I intended that as a description of the reasoning behind the design choices and the behaviour of the APIs, not a literal instruction to copy the code :-( 

> Calling postgres.delete() is the common practice by other "type" APIs, such as:

Sure. Those tend to have the nested callback and try / catch block structures that are hard to achieve coverage on.

There are also [API implementations](https://github.com/folio-org/mod-inventory-storage/blob/master/src/main/java/org/folio/rest/impl/BoundWithPartAPI.java) that use the PgUtil class when the intended behaviour is the typical baseline behaviour of the operation.

[the remark this answers is not in the document]
Comment (kaladay) [reply src/main/java/org/folio/rest/impl/RelatedInstanceTypeAPI.java]: Using PgUtil results in a much cleaner approach and I have switched to that.

**Review by marcjohnson-kint — APPROVED** [review]

**Pull request merged**